### PR TITLE
Update: Move Metric Parameter Config to Request Preview Column Config

### PR DESCRIPTION
**Pull request by mattgraham1995**
## Description
Metric column config component should be able to change the parameter for a given metric column.

## Proposed Changes

- Add UPDATE_METRIC_PARAM method to metric fragment and request action consumer
- Move fetching of metric parameter values into the metric-parameter service
- Change `editingColumn` property to depend on index of the column instead of the `canonicalName` since that will change with param changes. This is a temporary solution until a rework in the request modeling is done.
- Add metric column config component with column name and metric parameter options
- Add components for dimension and date time as well that export the base column config component for now

## Screenshots
Screen Shot 2019-12-10 at 5 56 00 PM

## License

I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: packages/data/addon/services/metric-parameter.js [hunk 1]
```diff
@@ -49,6 +49,42 @@ export default Service.extend({
     return this._supportedHandlers[meta.type](meta);
   },
 
+  /**
+   * @method fetchAllParams
+   * @param {Object} metricMeta - Metric metadata object
+   * @returns {Promise} promise hash with all the kinds of params and their values
+   */
+  fetchAllParams(metricMeta) {
```
Comment (mattgraham1995) [inline packages/data/addon/services/metric-parameter.js]: This method was changed a bit and moved here from the `metric-config` component

File: packages/reports/addon/components/navi-request-column-config/metric.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    this._previousMetric = null;
+    this._parametersPromise = null;
+  }
+
+  /**
+   * @override
+   * @method didReceiveAttrs
+   */
+  didReceiveAttrs() {
+    super.didUpdateAttrs(...arguments);
```
Comment (daniellevinson) [inline packages/reports/addon/components/navi-request-column-config/metric.js]: `this._super(...arguments);` ?

Comment (kevinhinterlong) [reply packages/reports/addon/components/navi-request-column-config/metric.js]: Nope this is the right way for native classes, but it was the wrong method 😮

File: packages/reports/addon/components/navi-request-column-config/metric.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    if (!metric.hasParameters) {
+      this.set('allParameters', null);
+      return;
+    }
+
+    set(
+      this,
+      'allParameters',
+      this.parameterService.fetchAllParams(metric).then(result => {
+        return Object.values(result);
```
[the remark this answers is not in the document]
Comment (mattgraham1995) [reply packages/reports/addon/components/navi-request-column-config/metric.js]: Blame prettier

File: packages/reports/addon/components/navi-request-preview.js [hunk 4]
```diff
@@ -45,17 +45,33 @@ class NaviRequestPreview extends Component {
...
   /**
    * @property {Object} editingColumn - The column object that is currently being edited, null if not editing
    */
-  editingColumn = null;
+  @computed('_editingColumnIndex', 'columns.[]')
+  get editingColumn() {
+    const columnIndex = this._editingColumnIndex;
+    const columns = this.columns;
+
+    if (typeof columnIndex === 'number' && columnIndex > -1 && columnIndex < this.columns.length) {
```
Comment (kevinhinterlong) [inline packages/reports/addon/components/navi-request-preview.js]:
```suggestion
    if (typeof columnIndex === 'number' && 0 <= columnIndex && columnIndex < this.columns.length) {
```

File: packages/reports/addon/consumers/request/metric.js [hunk 5]
```diff
@@ -42,6 +42,18 @@ export default ActionConsumer.extend({
       get(currentModel, 'request').addRequestMetricWithParam(metric, parameters);
     },
 
+    /**
+     * @action UPDATE_METRIC_PARAM
+     * @param {Object} route - route that has a model that contains a request property
+     * @param {String} metric - canonical name of metric to add
+     * @param {Object} parameter - selected metric parameter
```
Comment (daniellevinson) [inline packages/reports/addon/consumers/request/metric.js]: `parameterId, parameterKey`

File: packages/reports/addon/components/navi-request-preview.js [hunk 6]
```diff
@@ -45,17 +45,33 @@ class NaviRequestPreview extends Component {
...
+   */
+  _editingColumnIndex = null;
+
   /**
    * @property {Object} editingColumn - The column object that is currently being edited, null if not editing
    */
-  editingColumn = null;
+  @computed('_editingColumnIndex', 'columns.[]')
+  get editingColumn() {
+    const columnIndex = this._editingColumnIndex;
```
Comment (kevinhinterlong) [inline packages/reports/addon/components/navi-request-preview.js]: `... = this`?

File: packages/core/addon/models/bard-request/fragments/metric.js [hunk 7]
```diff
@@ -36,6 +36,18 @@ export default Fragment.extend(Copyable, Validations, {
     });
   }),
 
+  /**
+   * @method updateParameter - Update a parameter of the metric
+   * @param {String} parameterId - id property of parameter metadata object
+   * @param {String} parameterKey - name of the parameter type (e.g. currency)
```
Comment (daniellevinson) [inline packages/core/addon/models/bard-request/fragments/metric.js]: `paramId, paramKey`

File: packages/reports/addon/templates/components/navi-request-column-config/metric.hbs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  <div class="navi-request-column-config__body">
+    <div class="navi-request-column-config__option navi-request-column-config__column-name">
+      <label for="columnName" class="navi-request-column-config__column-name-label navi-request-column-config__option-label">Column Name</label>
+      <Input 
+        class="navi-request-column-config__column-name-input navi-request-column-config__option-input"
+        @id="columnName" 
+        @value={{@column.displayName}}
+        @key-up={{action @onUpdateColumnName}}
+      />
+    </div>
```
Comment (daniellevinson) [inline packages/reports/addon/templates/components/navi-request-column-config/metric.hbs]: can we keep the above in base or some wrapping component and use for dimension/date-time?

File: packages/reports/addon/components/navi-request-column-config/metric.js (hunk 2 above)
Comment (daniellevinson) [reply packages/reports/addon/components/navi-request-column-config/metric.js]: lol yeah I updated the comment a second later

File: packages/reports/tests/integration/components/navi-request-column-config/metric-test.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      name: metric.canonicalName,
+      displayName: 'Revenue',
+      sort: 'none',
+      fragment: metric
+    });
+    this.set('onClose', () => {
+      assert.ok(true, 'onClose called');
+    });
+    this.set('onUpdateColumnName', newName => {
+      assert.equal(newName, 'Money', 'New display name is passed to name update action');
```
Comment (kevinhinterlong) [inline packages/reports/tests/integration/components/navi-request-column-config/metric-test.js]: Did this one get skipped?

File: packages/reports/tests/integration/components/navi-request-preview-test.js [hunk 10]
```diff
@@ -183,5 +206,52 @@ module('Integration | Component | navi-request-preview', function(hooks) {
...
+
+    await selectChoose('.navi-request-column-config__parameter-trigger', 'Dollars (CAD)');
+    assert.dom('#columnName').hasValue('Revenue (CAD)', 'Column name updates with metric param');
+    assert.deepEqual(
+      textContentArray('.navi-request-preview__column-header'),
+      ['Date', 'Age', 'Banner Ad Clicks', 'Ad Clicks (VideoAds)', 'Revenue (CAD)'],
+      'Column name updates in column headers'
+    );
+
+    await fillIn('#columnName', "Dolla Dolla Bill Y'all");
```
Comment (kevinhinterlong) [inline packages/reports/tests/integration/components/navi-request-preview-test.js]: Maybe use [`typeIn`](https://github.com/emberjs/ember-test-helpers/blob/master/API.md#typein) in the future?

File: packages/reports/addon/consumers/request/metric.js [hunk 11]
```diff
@@ -46,7 +46,8 @@ export default ActionConsumer.extend({
      * @action UPDATE_METRIC_PARAM
      * @param {Object} route - route that has a model that contains a request property
      * @param {String} metric - canonical name of metric to add
-     * @param {Object} parameter - selected metric parameter
+     * @param {String} parameterId - id property of parameter object, value of param
```
Comment (kevinhinterlong) [inline packages/reports/addon/consumers/request/metric.js]: ©️  in this file

File: packages/core/addon/models/bard-request/fragments/metric.js [hunk 12]
```diff
@@ -38,8 +38,8 @@ export default Fragment.extend(Copyable, Validations, {
 
   /**
    * @method updateParameter - Update a parameter of the metric
-   * @param {String} parameterId - id property of parameter metadata object
-   * @param {String} parameterKey - name of the parameter type (e.g. currency)
+   * @param {String} paramId - id property of parameter metadata object
```
Comment (kevinhinterlong) [inline packages/core/addon/models/bard-request/fragments/metric.js]: ©️ here too

File: packages/reports/addon/templates/components/navi-request-column-config/metric.hbs (hunk 8 above)
Comment (mattgraham1995) [reply packages/reports/addon/templates/components/navi-request-column-config/metric.hbs]: Done 👍

**Pull request merged**